tests/distributed/_test_distributed.py

```diff
@@ -0,0 +1,156 @@
+        self.n_workers = n_workers
+
+    def worker_train(self, i: int) -> None:
+        """Start the training process on the `i`-th worker.
+
+        If this is the first worker, its logs are piped to stdout.
+        """
+        cmd = f'./lightgbm config=train{i}.conf'.split()
+        if i == 0:
+            return run_and_log(cmd)
+        subprocess.run(cmd)
+
+    def _set_ports(self) -> None:
+        """Randomly assign a port for training to each worker and save all ports to mlist.txt."""
+        self.listen_ports = [lgb.dask._find_random_open_port() for _ in range(self.n_workers)]
+        with open('mlist.txt', 'wt') as f:
+            for port in self.listen_ports:
+                f.write(f'127.0.0.1 {port}\n')
+
+    def _write_data(self, partitions: List[np.ndarray]) -> None:
+        """Write all training data as train.txt and each training partition as train{i}.txt."""
+        all_data = np.vstack(partitions)
+        np.savetxt('train.txt', all_data, delimiter=',')
+        for i, partition in enumerate(partitions):
+            np.savetxt(f'train{i}.txt', partition, delimiter=',')
+
+    def fit(self, partitions: List[np.ndarray]) -> None:
+        """Run the distributed training process on a single machine.
+
+        For each worker i:
+            1. The i-th partition is saved as train{i}.txt
+            2. A random port is assigned for training.
+            3. A configuration file train{i}.conf is created.
+            4. The lightgbm binary is called with config=train{i}.conf in another thread.
+        The whole training set is saved as train.txt and the logs from the first worker are piped to stdout.
+        """
+        self._write_data(partitions)
+        self.label_ = np.hstack([partition[:, 0] for partition in partitions])
+        self._set_ports()
+        futures = []
+        with ThreadPoolExecutor(max_workers=self.n_workers) as executor:
+            for i in range(self.n_workers):
+                self.write_train_config(i)
+                futures.append(executor.submit(self.worker_train, i))
+            _ = [f.result() for f in futures]
+
+    def predict(self) -> np.ndarray:
+        """Compute the predictions using the model created in the fit step.
+
+        model.txt is used to predict the training set train.txt using predict.conf.
+        The predictions are saved as predictions.txt and are then loaded to return them as a numpy array.
+        The logs are piped to stdout.
+        """
+        cmd = './lightgbm config=predict.conf'.split()
+        run_and_log(cmd)
+        y_pred = np.loadtxt('predictions.txt')
+        return y_pred
+
+    def write_train_config(self, i: int) -> None:
+        """Create a file train{i}.txt with the required configuration to train.
+
+        Each worker gets a different port and piece of the data, the rest are the
+        model parameters contained in `self.config`.
+        """
+        with open(f'train{i}.conf', 'wt') as f:
```

Comment:
I think they both need the same file and the data gets partitioned internally, I see this using the current approach:
```
[LightGBM] [Info] Number of data points in the train set: 248, number of used features: 4
[LightGBM] [Info] Number of data points in the train set: 252, number of used features: 4
```
Whereas using `lightgbm.dask`
```
[LightGBM] [Info] Number of data points in the train set: 500, number of used features: 4
[LightGBM] [Info] Number of data points in the train set: 500, number of used features: 4
```

Not sure if there's a way to tell each machine to use all of its data, I think right now it's splitting the data of the main machine.

Reply:
Ah: https://lightgbm.readthedocs.io/en/latest/Parameters.html#pre_partition

Reply:
haha ah I JUST found that too. Yeah, I think that is necessary for the CLI. I just tried adding `pre_partition: True` to the config and found that the returned model used all 1000 data points.

Ok, I think doing that should be enough. It would still probably be good to write two separate model files, to avoid any issues that might be caused by both processes trying to write to the same file.

Reply:
Adressed in 9f51ad8